lib/duniverse.ml

```diff
@@ -1,15 +1,148 @@
-  pins : Types.Opam.pin list;
-  packages : Types.Opam.entry list;
-  repos : Types.Dune.repo list;
-  remotes : string list; [@default []]
-  branch : string [@default "master"]
-}
-[@@deriving sexp]
+module Deps = struct
+  module Opam = struct
+    type t = { name : string; version : string option [@default None] [@sexp_drop_default] }
+    [@@deriving sexp]
+
+    let equal t t' = String.equal t.name t'.name && Option.equal String.equal t.version t'.version
+
+    let pp fmt = function
+      | { name; version = None } -> Format.fprintf fmt "%s" name
+      | { name; version = Some v } -> Format.fprintf fmt "%s.%s" name v
+
+    let raw_pp fmt { name; version } =
+      let open Pp_combinators.Ocaml in
+      Format.fprintf fmt "@[<hov 2>{ name = %S;@ version = %a }@]" name (option string) version
+  end
+
+  module Source = struct
+    module Package = struct
+      type t = { opam : Opam.t; upstream : string; ref : string }
+
+      let equal t t' =
+        Opam.equal t.opam t'.opam
+        && String.equal t.upstream t'.upstream
+        && String.equal t.ref t'.ref
+
+      let raw_pp fmt { opam; upstream; ref } =
+        Format.fprintf fmt "@[<hov 2>{ opam = %a;@ upstream = %S;@ ref = %S }@]" Opam.raw_pp opam
+          upstream ref
+    end
+
+    type t = {
+      dir : string;
+      upstream : string;
+      ref : string; [@default "master"] [@sexp_drop_default]
+      provided_packages : Opam.t list [@default []] [@sexp_drop_default]
+    }
+    [@@deriving sexp]
+
+    let equal t t' =
+      String.equal t.dir t'.dir && String.equal t.upstream t'.upstream && String.equal t.ref t'.ref
+
+    let raw_pp fmt { dir; upstream; ref; provided_packages } =
+      let open Pp_combinators.Ocaml in
+      Format.fprintf fmt
+        "@[<hov 2>{ dir = %S;@ upstream = %S;@ ref = %S;@ provided_packages = %a }@]" dir upstream
+        ref (list Opam.raw_pp) provided_packages
+
+    let from_package { Package.opam; upstream; ref } =
+      { dir = opam.name; upstream; ref; provided_packages = [ opam ] }
+
+    let aggregate t package =
+      let package_name = package.Package.opam.name in
+      let new_dir =
+        match String.compare t.dir package_name with Lt | Eq -> t.dir | Gt -> package_name
+      in
+      let new_ref =
+        match Ordering.of_int (OpamVersionCompare.compare t.ref package.ref) with
+        | Gt | Eq -> t.ref
+        | Lt -> package.ref
+      in
+      { t with
+        dir = new_dir;
+        ref = new_ref;
+        provided_packages = package.opam :: t.provided_packages
+      }
+
+    let aggregate_packages l =
+      let update map ({ Package.upstream; _ } as package) =
+        String.Map.update map upstream ~f:(function
+          | None -> Some (from_package package)
+          | Some t -> Some (aggregate t package) )
+      in
+      let aggregated_map = List.fold_left ~init:String.Map.empty ~f:update l in
+      String.Map.values aggregated_map
+  end
+
+  module One = struct
```

Comment:
Why `One` ? What about a more explicit name like `Classify` ?

Reply:
Yeah `One` is a really bad name, I'll think of a better one!

Reply:
`Classified` is great.